networkinterfaces: support source-directory statement

Carbenium opened the pull request:
This patch series implements support for the `source-directory` statement according to `interfaces(5)`. Additionally relative path handling in `source` and `source-directory` statements is fixed to match the documented behavior:


> When sourcing files or directories, if a path doesn't have a leading slash, it's considered relative to the directory containing the file in which the keyword is placed. In the example above, if the file is located at /etc/network/interfaces, paths to the included files are understood to be under /etc/network.

Excerpt from https://manpages.debian.org/buster/ifupdown/interfaces.5.en.html

Fixes #191

julienfortin commented:
Hi, I will review/approve the PR once I get access to my github account back (I lost my phone so I need to reset 2FA etc).

sbraz commented:
Hi @julienfortin, have you had time to check this PR yet?

julienfortin commented on ifupdown2/ifupdown/networkinterfaces.py:
```diff
@@ -176,6 +181,25 @@ def process_source(self, lines, cur_idx, lineno):
+
+            folders = glob.glob(sourced_directory)
+            for folder in folders:
+                filenames = [file for file in os.listdir(folder) if re.match(r'^[a-zA-Z0-9_-]+$', file)]
```
Why do we need a name check here? 
The following doesn't work with this regex check:
```
➜  ~ cat /etc/network/interfaces             
auto eth0
iface eth0 inet dhcp

source-directory interfaces.d
➜  ~ ll /etc/network/interfaces.d/
total 4.0K
-rw-r--r-- 1 root root 45 Dec 17 03:19 test2.intf
➜  ~ cat /etc/network/interfaces.d/test2.intf
auto test2
iface test2
      link-type dummy
➜  ~ 
➜  ~ ifquery -a
auto eth0
iface eth0 inet dhcp

➜  ~ 
```

Wouldn't:
```
+            for folder in glob.glob(sourced_directory):
+                for f in os.listdir(folder) or []:
+                    self.read_file(os.path.join(folder, f))
```
work just as well?

sbraz replied:
That is expected to match ifupdown's behaviour, see https://salsa.debian.org/debian/ifupdown/-/blob/0.8.36/interfaces.5.pre#L119-125

Carbenium replied:
I have nothing to add to @sbraz's explanation. That was exactly my intention.

julienfortin replied:
I think we can extend ifupdown's behavior and drop the regex check to cover a broader range of file names (like the one I used: ".intf").

What do you guys think?

julienfortin commented:
@sbraz Thanks for the reminder!

@Carbenium please review the requested change or please explain why the regex is needed

sbraz commented:
In my opinion, it all depends on what you're aiming for.
If the package is supposed to be a drop-in replacement for ifupdown, I wouldn't recommend it as it could lead to unintended behaviours. For instance, what if I created a file whose name was not valid, that file contained broken config, ifupdown never picked it up so I forgot about it ; then I install ifupdown2 and it breaks my networking?

However, as you already support new features, maybe it is your intention that there are breaking changes from ifupdown.

julienfortin commented:
Yes one of the goal is to replace ifupdown but also adding new features and improvements are really important to us. We don't want to limit our users to specific names (pretty sure i've seen Cumulus Linux users use the .intf format before).

@Carbenium unless there's a security reason or anything else, please update the PR to remove the regex and i'll merge it :)

Carbenium commented:
@julienfortin Sorry for the delay!

I implemented the discussed code changes and rebased the PR on current master. If you want me to squash the last commit in the series before merging, I'm happy to do so.

julienfortin commented:
@Carbenium no worries, thanks for making the change! :)
